File: java/BinarySearch/Find_K_Closest_Elements_658.java
package java.BinarySearch;

import java.util.*;
import java.io.*;

public class Find_K_Closest_Elements_658 {
    public List<Integer> findClosestElements(int[] arr, int k, int x) {
        int l, r, mid, temp = 0;
        l = 0;
        r = arr.length - 1;
        if (k > r + 1)
            return new ArrayList<>();
        while (l <= r) {
            mid = (l + r) / 2;
            if (arr[mid] == x) {
                temp = mid;
                break;
            }
            if (x < arr[mid])
                r = mid - 1;
            else
                l = mid + 1;
            if (Math.abs(arr[temp] - x) > Math.abs(arr[mid] - x))
                temp = mid;
        }
        l = temp - 1;
        r = temp + 1;
        List<Integer> ans = new ArrayList<Integer>();

        ans.add(arr[temp]);
        while ((l >= 0 || r < arr.length) && ans.size() < k) {
            if (l >= 0 && r < arr.length && (x - arr[l]) <= (arr[r] - x))
                ans.add(arr[l--]);
            else if (l >= 0 && r < arr.length && (x - arr[l]) > (arr[r] - x))
                ans.add(arr[r++]);
            else if (l >= 0)
                ans.add(arr[l--]);
            else {
                ans.add(arr[r++]);
            }
        }

        Collections.sort(ans);
        return ans;
    }

    public static void main(String[] args) throws FileNotFoundException {
        File f = new File("output.txt");

        // Scanner ip = new Scanner(System.in);
        try (Scanner ip = new Scanner(new File("input.txt"));
                PrintWriter op = new PrintWriter(f);) {

            // int x = ip.nextInt();
            // int y = ip.nextInt();
            // op.println(x + y);
            int grid[] = { 1, 2, 3, 4, 5 };
            Find_K_Closest_Elements_658 ob = new Find_K_Closest_Elements_658();
            List<Integer> ans = ob.findClosestElements(grid, 4, 3);
            op.println(ans);
        }
    }
}

File: Code/BinarySearch/Find_K_Closest_Elements_658.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution
{
public:
    vector<int> findClosestElements(vector<int> arr, int k, int x)
    {
        int l, r, mid, temp = 0;
        l = 0, r = arr.size() - 1;
        if (k > r + 1)
            return {};
        while (l <= r)
        {
            mid = (l + r) / 2;
            if (arr[mid] == x)
            {
                temp = mid;
                break;
            }
            if (x < arr[mid])
                r = mid - 1;
            else
                l = mid + 1;
            if (abs(arr[temp] - x) > abs(arr[mid] - x))
                temp = mid;
        }
        l = temp - 1;
        r = temp + 1;
        vector<int> ans;

        ans.push_back(arr[temp]);
        while ((l >= 0 || r < arr.size()) && ans.size() < k)
        {
            if (l >= 0 && r < arr.size() && (x - arr[l]) <= (arr[r] - x))
                ans.push_back(arr[l--]);
            else if (l >= 0 && r < arr.size() && (x - arr[l]) > (arr[r] - x))
                ans.push_back(arr[r++]);
            else if (l >= 0)
                ans.push_back(arr[l--]);
            else
            {
                ans.push_back(arr[r++]);
            }
        }

        sort(ans.begin(), ans.end());
        return ans;
    }
};
int main()
{
#ifndef ONLINE_JUDGE
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);
#endif
    Solution obj;
    auto t = obj.findClosestElements({1, 2, 3, 4, 5}, 4, 3);
    for (int i : t)
        cout << i << " ";
    return 0;
}